From d3bfb2e9987cff5346fc0d56ad1616c367a40b7c Mon Sep 17 00:00:00 2001
From: Andreas Blaesius <skate4life@gmx.de>
Date: Fri, 10 Mar 2017 12:00:16 +0100
Subject: [PATCH] espresso: Introduce an option to force a espresso variant

- On some espresso devices board detection does not work correctly after
  screen or mainboard got replaced. Give those people a chance too.

- Also add variant specific defconfig:
  On Android:
    TARGET_KERNEL_VARIANT_CONFIG := p****_defconfig
    and
    TARGET_KERNEL_CONFIG := espresso_defconfig
    or
    TARGET_KERNEL_CONFIG := espresso_kitkat_defconfig

  Compiling outside Android:
    make VARIANT_DEFCONFIG=p****_defconfig espresso_defconfig
    or
    make VARIANT_DEFCONFIG=p****_defconfig espresso_kitkat_defconfig

Change-Id: I4030324d734fce26a8e6ca0d5acb779a2afda681
---
 arch/arm/configs/p3100_defconfig     |  2 ++
 arch/arm/configs/p3110_defconfig     |  2 ++
 arch/arm/configs/p3113_defconfig     |  2 ++
 arch/arm/configs/p5100_defconfig     |  2 ++
 arch/arm/configs/p5110_defconfig     |  2 ++
 arch/arm/configs/p5113_defconfig     |  2 ++
 arch/arm/mach-omap2/Kconfig          | 41 ++++++++++++++++++++++++++++
 arch/arm/mach-omap2/board-espresso.c | 35 ++++++++++++++++++++++++
 8 files changed, 88 insertions(+)
 create mode 100644 arch/arm/configs/p3100_defconfig
 create mode 100644 arch/arm/configs/p3110_defconfig
 create mode 100644 arch/arm/configs/p3113_defconfig
 create mode 100644 arch/arm/configs/p5100_defconfig
 create mode 100644 arch/arm/configs/p5110_defconfig
 create mode 100644 arch/arm/configs/p5113_defconfig

diff --git a/arch/arm/configs/p3100_defconfig b/arch/arm/configs/p3100_defconfig
new file mode 100644
index 00000000000..af85b7b39ba
--- /dev/null
+++ b/arch/arm/configs/p3100_defconfig
@@ -0,0 +1,2 @@
+CONFIG_OMAP_SKIP_BOARDDETECTION=y
+CONFIG_OMAP_P3100=y
diff --git a/arch/arm/configs/p3110_defconfig b/arch/arm/configs/p3110_defconfig
new file mode 100644
index 00000000000..ba6f3260fe1
--- /dev/null
+++ b/arch/arm/configs/p3110_defconfig
@@ -0,0 +1,2 @@
+CONFIG_OMAP_SKIP_BOARDDETECTION=y
+CONFIG_OMAP_P3110=y
diff --git a/arch/arm/configs/p3113_defconfig b/arch/arm/configs/p3113_defconfig
new file mode 100644
index 00000000000..8fbbf3896c3
--- /dev/null
+++ b/arch/arm/configs/p3113_defconfig
@@ -0,0 +1,2 @@
+CONFIG_OMAP_SKIP_BOARDDETECTION=y
+CONFIG_OMAP_P3113=y
diff --git a/arch/arm/configs/p5100_defconfig b/arch/arm/configs/p5100_defconfig
new file mode 100644
index 00000000000..71b61afe103
--- /dev/null
+++ b/arch/arm/configs/p5100_defconfig
@@ -0,0 +1,2 @@
+CONFIG_OMAP_SKIP_BOARDDETECTION=y
+CONFIG_OMAP_P5100=y
diff --git a/arch/arm/configs/p5110_defconfig b/arch/arm/configs/p5110_defconfig
new file mode 100644
index 00000000000..d47fd6268b5
--- /dev/null
+++ b/arch/arm/configs/p5110_defconfig
@@ -0,0 +1,2 @@
+CONFIG_OMAP_SKIP_BOARDDETECTION=y
+CONFIG_OMAP_P5110=y
diff --git a/arch/arm/configs/p5113_defconfig b/arch/arm/configs/p5113_defconfig
new file mode 100644
index 00000000000..b96f8e45757
--- /dev/null
+++ b/arch/arm/configs/p5113_defconfig
@@ -0,0 +1,2 @@
+CONFIG_OMAP_SKIP_BOARDDETECTION=y
+CONFIG_OMAP_P5113=y
diff --git a/arch/arm/mach-omap2/Kconfig b/arch/arm/mach-omap2/Kconfig
index 8341d3756b3..6f950d80ea3 100644
--- a/arch/arm/mach-omap2/Kconfig
+++ b/arch/arm/mach-omap2/Kconfig
@@ -383,6 +383,47 @@ config MACH_OMAP4_ESPRESSO
 	select REGULATOR_FIXED_VOLTAGE
 	select TWL6030_GPADC
 
+config OMAP_SKIP_BOARDDETECTION
+	bool "Skip intelligent Espresso / Espresso10 detection"
+	depends on MACH_OMAP4_ESPRESSO
+	default n
+	help
+	  Select this if you want to force a Espresso / Espresso10 specific
+	  configuartion. Skipping intelligent board detection is only needed
+	  if your variant is not detected correctly (e.g. afer replacing
+	  the Display).
+
+choice
+	prompt "Select Espresso / Espresso10 Board"
+	depends on OMAP_SKIP_BOARDDETECTION
+	default OMAP_P5100
+
+config OMAP_P3100
+	bool "Force P3100 Tab 2 variant"
+	depends on OMAP_SKIP_BOARDDETECTION
+
+config OMAP_P3110
+	bool "Force P3110 Tab 2 variant"
+	depends on OMAP_SKIP_BOARDDETECTION
+
+config OMAP_P3113
+	bool "Force P3113 Tab 2 variant"
+	depends on OMAP_SKIP_BOARDDETECTION
+
+config OMAP_P5100
+	bool "Force P5100 Tab 2 variant"
+	depends on OMAP_SKIP_BOARDDETECTION
+
+config OMAP_P5110
+	bool "Force P5110 Tab 2 variant"
+	depends on OMAP_SKIP_BOARDDETECTION
+
+config OMAP_P5113
+	bool "Force P5113 Tab 2 variant"
+	depends on OMAP_SKIP_BOARDDETECTION
+
+endchoice
+
 config MACH_TUNA
 	bool "Tuna Board"
 	default n
diff --git a/arch/arm/mach-omap2/board-espresso.c b/arch/arm/mach-omap2/board-espresso.c
index ae119612e10..341bbfd4791 100644
--- a/arch/arm/mach-omap2/board-espresso.c
+++ b/arch/arm/mach-omap2/board-espresso.c
@@ -110,9 +110,40 @@ static struct omap_musb_board_data musb_board_data = {
 
 /* Board identification */
 
+#ifdef CONFIG_OMAP_SKIP_BOARDDETECTION
+#if defined(CONFIG_OMAP_P3100)
 static bool _board_has_modem = true;
+static bool _board_is_espresso10 = false;
+static bool _board_is_bestbuy_variant = false;
+#elif defined(CONFIG_OMAP_P3110)
+static bool _board_has_modem = false;
+static bool _board_is_espresso10 = false;
+static bool _board_is_bestbuy_variant = false;
+#elif defined(CONFIG_OMAP_P3113)
+static bool _board_has_modem = false;
+static bool _board_is_espresso10 = false;
+static bool _board_is_bestbuy_variant = true;
+#elif defined(CONFIG_OMAP_P5100)
+static bool _board_has_modem = true;
+static bool _board_is_espresso10 = true;
+static bool _board_is_bestbuy_variant = false;
+#elif defined(CONFIG_OMAP_P5110)
+static bool _board_has_modem = false;
 static bool _board_is_espresso10 = true;
 static bool _board_is_bestbuy_variant = false;
+#elif defined(CONFIG_OMAP_P5113)
+static bool _board_has_modem = false;
+static bool _board_is_espresso10 = true;
+static bool _board_is_bestbuy_variant = true;
+#endif
+
+#else
+
+static bool _board_has_modem = true;
+static bool _board_is_espresso10 = true;
+static bool _board_is_bestbuy_variant = false;
+
+#endif
 
 /*
  * Sets the board type
@@ -136,17 +167,20 @@ static __init int setup_board_type(char *str)
 		return 1;
 	}
 
+#ifndef CONFIG_OMAP_SKIP_BOARDDETECTION
 	/*
 	 * P51xx bootloaders pass lcd_id=1 and on some older lcd_id=0,
 	 * everything else is P31xx.
 	 */
 	if (lcd_id > 1)
 		_board_is_espresso10 = false;
+#endif
 
 	return 0;
 }
 early_param("lcd_panel_id", setup_board_type);
 
+#ifndef CONFIG_OMAP_SKIP_BOARDDETECTION
 /*
  * Sets whether the device is a wifi-only variant
  */
@@ -191,6 +225,7 @@ bool board_is_bestbuy_variant(void) {
 }
 
 /* Board identification end */
+#endif
 
 static void espresso_power_off_charger(void)
 {
